### PR TITLE
Adds `remove_if` Function to `linestring_intersection_intermediates`

### LINKED ISSUE #853: Create `remove_if` for `intersection_intermediates` to Remove Geometries Given a Flag Array

### PR DESCRIPTION
**Pull request by isVoid**
## Description

https://github.com/rapidsai/cuspatial/pull/813 adds a struct to hold the intermediate result from `linestring_intersection_with_duplicates`. After discovering the duplicates from the result, we need a way to remove these geometries from the struct, as well as updating the offsets and look-back ids. This PR adds a `remove_if` function to the struct, which models after `thrust::remove_if` that accepts a range of flags, and removes the geometries in the intermediates where the flag is set to 1.

Closes #853 
Depends on #813. Contributes to #765.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by harrism** [comment]
I reviewed the remove_if function and it looks fine. Could maybe use some more explanatory documentation of the algorithm since it's a fairly complicated function. Will review more fully once dependencies are merged.

**Review by harrism — APPROVED** [review]

**Comment by isVoid** [comment]
rerun tests

**Comment by isVoid** [comment]
@thomcom I believe I addressed all your reviews regarding `remove_if` you posted in #852. I added more comments per https://github.com/rapidsai/cuspatial/pull/852#discussion_r1048760084. As for https://github.com/rapidsai/cuspatial/pull/852#discussion_r1048718340, I believe the naming is a convention I pulled from thrust and is a common short hand for "remove element if the flag is set". This is similar to `copy_if`.

**Review by thomcom — APPROVED** [review]

File: cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh [hunk 1]
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
                            thrust::next(offsets->begin()));
   }
 
+  /// Given a flag array, remove the ith geometry if `flags[i] == 1`.
+  /// @pre flag array must have the same size as the geometry array.
+  template <typename FlagRange>
+  void remove_if(FlagRange flags, rmm::cuda_stream_view stream)
+  {
+    // The offsets for the geometry array marks the start index for each list.
+    // When geometry are removed, the offset for the next list must
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:204]:
```suggestion
    // When geometries are removed, the offset for the next list must
```

File: cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh [hunk 2]
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
...
+
+    auto [keys_end, flags_end] =
+      thrust::reduce_by_key(rmm::exec_policy(stream),
+                            keys_begin,
+                            keys_begin + flags.size(),
+                            flags.begin(),
+                            reduced_keys.begin(),
+                            reduced_flags.begin(),
+                            thrust::equal_to<index_t>(),
+                            thrust::plus<index_t>());  // explicitly cast flagss to index_t type
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:221]:
```suggestion
                            thrust::plus<index_t>());  // explicitly cast flags to index_t type
```

File: cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh [hunk 3]
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
...
+                                                       // before adding to avoid overflow.
+
+    reduced_keys.resize(thrust::distance(reduced_keys.begin(), keys_end), stream);
+    reduced_flags.resize(thrust::distance(reduced_flags.begin(), flags_end), stream);
+
+    // Use `inclusive_scan` to compute the number of removed geometry in *all* previous lists.
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:227]:
```suggestion
    // Use `inclusive_scan` to compute the number of removed geometries in *all* previous lists.
```

File: cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh [hunk 4]
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
...
+    thrust::transform(
+      rmm::exec_policy(stream),
+      offsets->begin(),
+      offsets->end(),
+      thrust::make_counting_iterator(0),
+      offsets->begin(),
+      intersection_functors::offsets_update_functor{
+        reduced_keys.begin(), reduced_keys.end(), reduced_flags.begin(), reduced_flags.end()});
+
+    // Remove the geometry and corresponding id array per flag.
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:241]:
```suggestion
    // Remove the geometries and the corresponding ids per flag.
```

**Review by isVoid — COMMENTED** [review]
A couple typos

**Comment by isVoid** [comment]
@gpucibot  merge

**Comment by isVoid** [comment]
@gpucibot merge

**Pull request merged**